### PR TITLE
tests: Migrate defrag tests to common framework

### PR DESCRIPTION
**Pull request by kkkkun**
Part of https://github.com/etcd-io/etcd/issues/13637

### CODE REVIEW

File: tests/common/compact_test.go [hunk 1]
```diff
@@ -46,11 +46,7 @@ func TestCompact(t *testing.T) {
 			defer clus.Close()
 			testutils.ExecuteWithTimeout(t, 10*time.Second, func() {
 				var kvs = []testutils.KV{{Key: "key", Val: "val1"}, {Key: "key", Val: "val2"}, {Key: "key", Val: "val3"}}
-				for i := range kvs {
-					if err := clus.Client().Put(kvs[i].Key, kvs[i].Val); err != nil {
-						t.Fatalf("compactTest #%d: put kv error (%v)", i, err)
-					}
-				}
+				maintenanceInitKeys(t, clus.Client(), kvs...)
```
Comment (serathius) [inline tests/common/compact_test.go:49]: I have reservations about creating a helper function here. I think as long this is a single loop with calling put, it's more clearer than function call. Proposed name `maintenanceInitKeys` doesn't really make it obvious what it does. We could fix the name, however I would argue that this code is simple enough that it's not worth to do so.

Can you just revert this change?

File: tests/common/defrag_test.go [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+	"testing"
+	"time"
+
+	"go.etcd.io/etcd/tests/v3/framework"
+
+	"go.etcd.io/etcd/tests/v3/framework/config"
+	"go.etcd.io/etcd/tests/v3/framework/testutils"
+)
+
+func TestDefragOnline(t *testing.T) {
```
Comment (serathius) [inline tests/common/defrag_test.go:27]: From what I see TestDefragOnline and TestDefragOffline differ only by options?
Could we have table test `TestDefrag` with 2 subtests `Online` and `Offline`?

File: tests/common/compact_test.go (hunk 1 above)
Comment (kkkkun) [reply tests/common/compact_test.go:49]: Ok.

File: tests/e2e/ctl_v3_defrag_test.go [hunk 3]
```diff
@@ -1,69 +0,0 @@
...
-
-import (
-	"testing"
-
-	"go.etcd.io/etcd/tests/v3/framework/e2e"
-)
-
-func TestCtlV3DefragOnline(t *testing.T) { testCtl(t, defragOnlineTest) }
-
-func TestCtlV3DefragOfflineEtcdutl(t *testing.T) {
```
Comment (serathius) [inline tests/e2e/ctl_v3_defrag_test.go:25]: I don't think we should migrate `TestCtlV3DefragOfflineEtcdutl` to common framework. It tests specifically `etcdutl` binary, that is not used in integration tests. You might have noticed that current framework doesn't really handle it well, which can be seen it additional changes were required to make it work. 

In the first pass of the migration we should migrate only tests for features available in API, as they should work exactly the same for etcd client used in integration tests and binaries used in e2e binaries.

File: tests/framework/config/client.go [hunk 4]
```diff
@@ -41,3 +42,8 @@ type CompactOption struct {
 	Physical bool
 	Timeout  time.Duration
 }
+
+type DefragOption struct {
+	Timeout time.Duration
+	WithUtl bool
```
Comment (serathius) [inline tests/framework/config/client.go:48]: Let's not add `WithUtl` to the common framework, as it currently cannot be implemented for integration tests.

File: tests/framework/e2e/etcdctl.go [hunk 5]
```diff
@@ -243,3 +249,24 @@ func (ctl *EtcdctlV3) Health() error {
 	return SpawnWithExpects(args, map[string]string{}, lines...)
 
 }
+
+func (ctl *EtcdctlV3) Defragment(o config.DefragOption) error {
+
+	args := append([]string{}, "defrag")
```
Comment (serathius) [inline tests/framework/e2e/etcdctl.go:255]:
```suggestion
	args := []string{"defrag"}
```

File: tests/framework/e2e/etcdctl.go [hunk 6]
```diff
@@ -243,3 +249,24 @@ func (ctl *EtcdctlV3) Health() error {
...
+	if o.WithUtl {
+		args = append(ctl.amdArgsWithUtl(), args...)
+		args = append(args, "--data-dir", ctl.runningCfg[0].DataDirPath)
+		lines := []string{"finished defragmenting directory"}
+		return SpawnWithExpects(args, map[string]string{}, lines...)
+	}
+	args = append(ctl.cmdArgs(), args...)
+	lines := make([]string, len(ctl.endpoints))
+	for i := range lines {
+		lines[i] = "Finished defragmenting etcd member"
```
Comment (serathius) [inline tests/framework/e2e/etcdctl.go:268]: Based on this I surmise that etcdctl will send defrag request to each etcd member?

File: tests/common/defrag_test.go (hunk 2 above)
Comment (kkkkun) [reply tests/common/defrag_test.go:27]: If not migrating `etcdutl` to common, it will keep online. Offline would be delete.

File: tests/framework/e2e/etcdctl.go (hunk 6 above)
Comment (kkkkun) [reply tests/framework/e2e/etcdctl.go:268]: It will send to etcd member with `--endpoint`. 

> Based on this I surmise that etcdctl will send defrag request to each etcd member?


In current. Yes

File: tests/framework/e2e/etcdctl.go [hunk 7]
```diff
@@ -24,14 +24,16 @@ import (
...
 }
 
-func NewEtcdctl(cfg *EtcdProcessClusterConfig, endpoints []string) *EtcdctlV3 {
+func NewEtcdctl(cfg *EtcdProcessClusterConfig, endpoints []string, runningCfg []*EtcdServerProcessConfig) *EtcdctlV3 {
 	return &EtcdctlV3{
-		cfg:       cfg,
-		endpoints: endpoints,
+		cfg:        cfg,
+		endpoints:  endpoints,
+		runningCfg: runningCfg,
```
Comment (serathius) [inline tests/framework/e2e/etcdctl.go:36]: I don't think `[]*EtcdServerProcessConfig` is needed in the EtcdctlV3 now.

Comment (kkkkun) [reply tests/framework/e2e/etcdctl.go:36]: You are right. it's not needed now.
I keep this for `etcdutl` in the furture.  I will delete it for  your suggestion.

Comment (serathius) [reply tests/framework/e2e/etcdctl.go:36]: I don't think we should add this code just in case we need it. We might take different approach and not use this code at all.
Let's just remove it.

Comment (kkkkun) [reply tests/framework/e2e/etcdctl.go]: Removed.

File: tests/e2e/ctl_v3_defrag_test.go [hunk 8]
```diff
@@ -35,16 +34,25 @@ func maintenanceInitKeys(cx ctlCtx) {
 	}
 }
 
-func defragOnlineTest(cx ctlCtx) {
-	maintenanceInitKeys(cx)
+func ctlV3OfflineDefrag(cx ctlCtx) error {
```
Comment (serathius) [inline tests/e2e/ctl_v3_defrag_test.go:37]: Let's avoid reordering the functions to avoid obscuring git history.

**Review by serathius — APPROVED** [review]
LGTM, will merge when tests pass.
